### PR TITLE
Datascience

### PR DESCRIPTION
**Pull request by shishirtewari**
Please review data science changes.

### CODE REVIEW

File: data_science/spec/fileparser_spec.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+load 'fileparser.rb'
+require 'spec_helper'
+require 'rspec'
+
+describe FileParser do
+  context 'raise an error when' do
```
Comment (zsyed1) [inline data_science/spec/fileparser_spec.rb]: Add a _happy_ path test: `expect { FileParser.parse_data('valid.json') }.not_to raise_error(RuntimeError)`

File: data_science/datascience.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  def conversion_miss_count(name)
+    @cohorts.count { |cohort| cohort.name == name && cohort.result == 0 }
+  end
+
+  def chisqaure_p
+    @analyzer_data = {}
+    @cohorts.uniq(&:name).each do |row|
+      @analyzer_data[row.name] ||=
```
Comment (zsyed1) [inline data_science/datascience.rb]: Since you are calling `uniq` you shouldn't need the `||=` correct?

File: data_science/datascience.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
+require 'ABAnalyzer'
+load 'fileparser.rb'
+
+class DataScience
+  attr_accessor :cohorts, :ab_analyzer, :analyzer_data
+
```
Comment (zsyed1) [inline data_science/datascience.rb]: In line 5, `ab_analyzer` can be removed. I don't think it's used in the code.

Comment (zsyed1) [inline data_science/datascience.rb]: In line 5, `analyzer_data` is not being used anywhere so you can remove it. Also see note below about the same variable name.

File: data_science/datascience.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    @cohorts.count { |cohort| cohort.name == name && cohort.result == 1 }
+  end
+
+  def conversion_miss_count(name)
+    @cohorts.count { |cohort| cohort.name == name && cohort.result == 0 }
+  end
+
+  def chisqaure_p
+    @analyzer_data = {}
+    @cohorts.uniq(&:name).each do |row|
```
Comment (zsyed1) [inline data_science/datascience.rb]: The variable `analyzer_data` is local to this method `chisquare_p`. So you don't need the `@` symbol in its name.

File: data_science/datascience.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  def conversion_count(name)
+    @cohorts.count { |cohort| cohort.name == name && cohort.result == 1 }
+  end
+
+  def conversion_miss_count(name)
+    @cohorts.count { |cohort| cohort.name == name && cohort.result == 0 }
+  end
+
+  def chisqaure_p
+    @analyzer_data = {}
```
Comment (zsyed1) [inline data_science/datascience.rb]: Spelling: `chisquare_p`

File: data_science/datascience.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    end
+    leader.max_by { |_k, v| v }[:name]
+  end
+
+  def distribution_limit(name)
+    r = conversion_rate(name)
+    eb = error_bar(name)
+    [(r - eb).round(4), (r + eb).round(4)]
+  end
+
```
Comment (zsyed1) [inline data_science/datascience.rb]: Nice `distribution_limit(name)`

File: data_science/datascience.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    @cohorts = FileParser.parse_data(filename)
+  end
+
+  def cohort_count(name)
+    @cohorts.count { |cohort| cohort.name == name }
+  end
+
+  def conversion_count(name)
+    @cohorts.count { |cohort| cohort.name == name && cohort.result == 1 }
+  end
```
Comment (zsyed1) [inline data_science/datascience.rb]: I link the way you've implemented this (line 16 and others like it). Just wanted to point out that another way to do this is to move the logic `cohort.result == 1` into the `Cohort` class and call it a method:

```
def conversion?
  @result == 1
end
```
Then here line 16 could change to: `@cohorts.count { |cohort| cohort.name == name && cohort.conversion?}`. The same goes with `conversion_miss_count` line 20.

File: data_science/cohort.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class Cohort
+  attr_accessor :name, :result, :date
+
+  def initialize(row)
+    @date = row['date']
+    @name = row['cohort']
+    @result = row['result']
+  end
+end
```
Comment (zsyed1) [inline data_science/cohort.rb]: Just FYI, your `Cohort` class really represents a `Visit` more than a `Cohort`. Here there are two cohorts: A and B. Each of them visits the website many times, once for each row.

File: data_science/spec/cohort_spec.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+load 'cohort.rb'
+require 'spec_helper'
+require 'rspec'
+
+describe Cohort do
+  cohort = {}
+  cohort['date'] = '2014-03-20'
+  cohort['cohort'] = 'B'
+  cohort['result'] = 0
+
```
Comment (zsyed1) [inline data_science/spec/cohort_spec.rb]: Probably want to put lines 6 through 9 into the `let` on line 11. I would also change the name from `cohortobj` to simply `cohort`.

File: data_science/spec/datascience_spec.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    it 'Should calculate distribution range for each cohort' do
+      expect(ds.distribution_limit('A')).to match_array([0.005, 0.2056])
+      expect(ds.distribution_limit('B')).to match_array([0.1896, 0.3022])
+    end
+    it 'Should calculate chisqaure in sample size' do
+      expect(ds.chisqaure_p).to eq(0.9058)
+    end
+    it 'Should check who is leader' do
+      expect(ds.leader).to eq('B')
+    end
```
Comment (zsyed1) [inline data_science/spec/datascience_spec.rb]: Nice test. Btw there is much debate about whether or not to use the word `Should` in tests.

File: data_science/datascience.rb (hunk 3 above)
Comment (zsyed1) [inline data_science/datascience.rb]: In line 5, `cohorts` is never used outside this class. So it doesn't have to be in `attr_accessor`.

File: data_science/spec/spec_helper.rb [hunk 11]
```diff
@@ -0,0 +1,3 @@
+require_relative '../fileparser.rb'
+require_relative '../cohort.rb'
+require_relative '../datascience.rb'
```
Comment (zsyed1) [inline data_science/spec/spec_helper.rb]: It was a pleasure reading your code. Nice work.

File: data_science/spec/datascience_spec.rb (hunk 10 above)
Comment (djkotowski) [inline data_science/spec/datascience_spec.rb]: I prefer not having `should` in descriptions.

Comment (djkotowski) [inline data_science/spec/datascience_spec.rb]: Also, given these lines describe methods, why not `describe` the methods?

e.g. `describe "#conversion_rate"`

**Comment by djkotowski** [comment]
This is looking good. Take a look at the notes and reactor. You should only need one resubmission on that.

**Pull request closed without merging**